// jsii-reflect/lib/type-ref.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeReference = void 0;
const jsii = __importStar(require("@jsii/spec"));
class TypeReference {
    constructor(system, spec) {
        this.system = system;
        this.spec = spec;
    }
    toString() {
        if (this.void) {
            return 'void';
        }
        if (this.primitive) {
            return this.primitive;
        }
        if (this.fqn) {
            return this.fqn;
        }
        if (this.arrayOfType) {
            return `Array<${this.arrayOfType.toString()}>`;
        }
        if (this.mapOfType) {
            return `Map<string => ${this.mapOfType.toString()}>`;
        }
        if (this.unionOfTypes) {
            const union = this.unionOfTypes.map((x) => x.toString());
            union.sort();
            return union.join(' | ');
        }
        if (this.intersectionOfTypes) {
            const inter = this.intersectionOfTypes.map((x) => x.toString());
            inter.sort();
            return inter.join(' & ');
        }
        throw new Error('Invalid type reference');
    }
    get void() {
        return !this.spec;
    }
    get isAny() {
        return this.primitive === 'any';
    }
    get primitive() {
        if (!jsii.isPrimitiveTypeReference(this.spec)) {
            return undefined;
        }
        return this.spec.primitive;
    }
    get fqn() {
        return jsii.isNamedTypeReference(this.spec) ? this.spec.fqn : undefined;
    }
    get type() {
        if (!jsii.isNamedTypeReference(this.spec)) {
            return undefined;
        }
        return this.system.findFqn(this.spec.fqn);
    }
    get arrayOfType() {
        if (!jsii.isCollectionTypeReference(this.spec)) {
            return undefined;
        }
        if (this.spec.collection.kind !== jsii.CollectionKind.Array) {
            return undefined;
        }
        return new TypeReference(this.system, this.spec.collection.elementtype);
    }
    get mapOfType() {
        if (!jsii.isCollectionTypeReference(this.spec)) {
            return undefined;
        }
        if (this.spec.collection.kind !== jsii.CollectionKind.Map) {
            return undefined;
        }
        return new TypeReference(this.system, this.spec.collection.elementtype);
    }
    get unionOfTypes() {
        if (!jsii.isUnionTypeReference(this.spec)) {
            return undefined;
        }
        return this.spec.union.types.map((t) => new TypeReference(this.system, t));
    }
    get intersectionOfTypes() {
        if (!jsii.isIntersectionTypeReference(this.spec)) {
            return undefined;
        }
        return this.spec.intersection.types.map((t) => new TypeReference(this.system, t));
    }
}
exports.TypeReference = TypeReference;
